tensorcircuit/abstractcircuit.py

```diff
@@ -0,0 +1,609 @@
+
+        :Example:
+        >>> c = tc.Circuit(3)
+        >>> c.H(1)
+        >>> c.X(2)
+        >>> c.CNOT(0, 1)
+        >>> c.draw(output='text')
+        q_0: ───────■──
+             ┌───┐┌─┴─┐
+        q_1: ┤ H ├┤ X ├
+             ├───┤└───┘
+        q_2: ┤ X ├─────
+             └───┘
+        """
+        return self.to_qiskit().draw(**kws)
+
+    @classmethod
+    def from_qiskit(
+        cls, qc: Any, n: Optional[int] = None, inputs: Optional[List[float]] = None
+    ) -> "AbstractCircuit":
+        """
+        Import Qiskit QuantumCircuit object as a ``tc.Circuit`` object.
+
+        :Example:
+
+        >>> from qiskit import QuantumCircuit
+        >>> qisc = QuantumCircuit(3)
+        >>> qisc.h(2)
+        >>> qisc.cswap(1, 2, 0)
+        >>> qisc.swap(0, 1)
+        >>> c = tc.Circuit.from_qiskit(qisc)
+
+        :param qc: Qiskit Circuit object
+        :type qc: QuantumCircuit in Qiskit
+        :param n: The number of qubits for the circuit
+        :type n: int
+        :param inputs: possible input wavefunction for ``tc.Circuit``, defaults to None
+        :type inputs: Optional[List[float]], optional
+        :return: The same circuit but as tensorcircuit object
+        :rtype: Circuit
+        """
+        from .translation import qiskit2tc
+
+        if n is None:
+            n = qc.num_qubits
+
+        return qiskit2tc(qc.data, n, inputs, is_dm=cls.is_dm)  # type: ignore
+
+    def vis_tex(self, **kws: Any) -> str:
+        """
+        Generate latex string based on quantikz latex package
+
+        :return: Latex string that can be directly compiled via, e.g. latexit
+        :rtype: str
+        """
+        if self.inputs is not None:
+            init = ["" for _ in range(self._nqubits)]
+            init[self._nqubits // 2] = "\psi"
+            okws = {"init": init}
+        else:
+            okws = {"init": None}  # type: ignore
+        okws.update(kws)
+        return qir2tex(self._qir, self._nqubits, **okws)  # type: ignore
+
+    tex = vis_tex
+
+    def select_gate(self, which: Tensor, kraus: Sequence[Gate], *index: int) -> None:
+        """
+        Apply ``which``-th gate from ``kraus`` list, i.e. apply kraus[which]
+
+        :param which: Tensor of shape [] and dtype int
+        :type which: Tensor
+        :param kraus: A list of gate in the form of ``tc.gate`` or Tensor
+        :type kraus: Sequence[Gate]
+        :param index: the qubit lines the gate applied on
+        :type index: int
+        """
+        kraus = [k.tensor if isinstance(k, tn.Node) else k for k in kraus]
+        kraus = [gates.array_to_tensor(k) for k in kraus]
+        l = len(kraus)
+        r = backend.onehot(which, l)
+        r = backend.cast(r, dtype=dtypestr)
+        tensor = reduce(add, [r[i] * kraus[i] for i in range(l)])
+        self.any(*index, unitary=tensor)  # type: ignore
+
+    conditional_gate = select_gate
+
+    def cond_measurement(self, index: int) -> Tensor:
```

Comment:
somehow not sure whether we should put this method in abstractclass or baseclass